Core/include/Acts/Geometry/detail/DefaultGeometryIdGetter.hpp

```diff
@@ -0,0 +1,45 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Geometry/GeometryID.hpp"
+
+namespace Acts {
+namespace detail {
+
+/// Uniform access to geometry identifiers for different object types.
+struct DefaultGeometryIdGetter {
+  // explicit geometry identifiers are just forwarded.
+  constexpr GeometryID operator()(GeometryID id) const { return id; }
+  // encoded identifiers are converted back to geometry identifiers.
+  constexpr GeometryID operator()(GeometryID::Value encoded) const {
+    return GeometryID(encoded);
+  }
+  // support objects that implement `.geoID()`.
+  template <typename object_t>
+  inline auto operator()(const object_t& object) const
+      -> decltype(object.geoID(), GeometryID()) {
```

Comment:
How does this work? Is this SFINAE? `decltype` only takes one argument according to the specification, doesn't it?

Reply:
Yes, this is SFINAE. In this variant the comma is the [C++ comma operator](https://en.cppreference.com/w/cpp/language/operator_other#Built-in_comma_operator) that executes both statements, i.e. `object.geoID()` and `GeometryID()`, but evaluates only to the value of the last operand (here an instance of `GeometryID()`. This means that the return type is `GeometryID` iff the input object has the `.geoID()` method.

I agree that this is a bit of magic, but it is by far the most concise way of checking for a specific member function. It does not involve any extra libraries, helpers, or type traits.

Reply:
What happens if `decltype(object.geoID()) != GeometryID`?

Reply:
If the return value of `object.geoID()` is not convertible to a `GeometryID` you get a compile error. `object.geoID()` is a valid expression so the return type of the function is `GeometryID`. Thus `return object.geoID()` must be convertible to `GeometryID` for the code to compile.

Reply:
Ok I understand how this works now I think. If a type `T` doesn't implement `geoID` or `geometryID` this is not a substitution failure. You get a compiler error that kind of hints at what's happening though, so it's probably alright. The case that the method doesn't return `GeometryID` is also not substitution failure, but the compiler error is equally acceptable. I guess 'proper' SFINAE would be more wordy, and probably doesn't result in much nicer error messages, so ok.

Reply:
You also get an error, btw, if a type implements **BOTH** `geoID()` **AND** `geometryId()`, even if one of them returns something other than a `GeometryID`.

Reply:
Good point.  Would you not get a compile error since then you define a method with the same signature twice?

Reply:
You get a compile error that the overloads are ambiguous. If both are implemented, the compiler doesn't get to the point where the incompatibility between the method and `GeometryID` throws a compile error.

Reply:
This is arguably a construed example, why would you have a type that has `GeometryID geoID() const;` and also `MyOwnGeometryID geometryId() const;`.

Reply:
You could **probably** get around this issue by using `std::enable_if` and 'proper' SFINAE in a series of `if constexpr` in a single overload, and then doing a conditional `static_assert` in the final `else` case, with a more informative error message.

Reply:
> This is arguably a construed example, why would you have a type that has `GeometryID geoID() const;` and also `MyOwnGeometryID geometryId() const;`. This would be, from my point of view, an argument for less magic, and forcing client code to provide their own getter.

Yes, very construed. No sane code should implement both. Let's count on the sanity of us developers and future users for now?